internal/manifestgen/manifestgen.go

```diff
@@ -0,0 +1,176 @@
+	for name, commitSources := range commitSources {
+		commitSpecs := make([]ostree.CommitSpec, len(commitSources))
+		for idx, commitSource := range commitSources {
+			var err error
+			commitSpecs[idx], err = ostree.Resolve(commitSource)
+			if err != nil {
+				return nil, err
+			}
+		}
+		commits[name] = commitSpecs
+	}
+	return commits, nil
+}
+
+type (
+	DepsolveFunc func(cacheDir string, packageSets map[string][]rpmmd.PackageSet, d distro.Distro, arch string) (map[string][]rpmmd.PackageSpec, map[string][]rpmmd.RepoConfig, error)
+
+	ContainerResolverFunc func(containerSources map[string][]container.SourceSpec, archName string) (map[string][]container.Spec, error)
+
+	CommitResolverFunc func(commitSources map[string][]ostree.SourceSpec) (map[string][]ostree.CommitSpec, error)
+)
+
+// Options contains the optional optons for the manifest generation.
+// For unset options defaults will be used.
+type Options struct {
+	Cachedir          string
+	Output            io.Writer
+	Depsolver         DepsolveFunc
+	ContainerResolver ContainerResolverFunc
+	CommitResolver    CommitResolverFunc
+}
+
+// Generator can generate an osbuild manifest from a given repository
+// and options.
+type Generator struct {
+	cacheDir string
+	out      io.Writer
+
+	depsolver         DepsolveFunc
+	containerResolver ContainerResolverFunc
+	commitResolver    CommitResolverFunc
+
+	reporegistry *reporegistry.RepoRegistry
+}
+
+// New will create a new manifest generator
+func New(reporegistry *reporegistry.RepoRegistry, opts *Options) (*Generator, error) {
+	if opts == nil {
+		opts = &Options{}
+	}
+	mg := &Generator{
+		reporegistry: reporegistry,
+
+		cacheDir:          opts.Cachedir,
+		out:               opts.Output,
+		depsolver:         opts.Depsolver,
+		containerResolver: opts.ContainerResolver,
+		commitResolver:    opts.CommitResolver,
+	}
+	if mg.out == nil {
+		mg.out = os.Stdout
+	}
+	if mg.depsolver == nil {
+		mg.depsolver = defaultDepsolver
+	}
+	if mg.containerResolver == nil {
+		mg.containerResolver = defaultContainerResolver
+	}
+	if mg.commitResolver == nil {
+		mg.commitResolver = defaultCommitResolver
+	}
+
+	return mg, nil
+}
+
+// Generate will generate a new manifest for the given distro/imageType/arch
+// combination.
+func (mg *Generator) Generate(bp *blueprint.Blueprint, dist distro.Distro, imgType distro.ImageType, a distro.Arch, imgOpts *distro.ImageOptions) error {
+	if imgOpts == nil {
+		imgOpts = &distro.ImageOptions{}
+	}
+
+	repos, err := mg.reporegistry.ReposByImageTypeName(dist.Name(), a.Name(), imgType.Name())
+	if err != nil {
+		return err
+	}
+	preManifest, warnings, err := imgType.Manifest(bp, *imgOpts, repos, 0)
+	if err != nil {
+		return err
+	}
+	if len(warnings) > 0 {
+		// XXX: what can we do here? for things like json output?
+		// what are these warnings?
+		return fmt.Errorf("warnings during manifest creation: %v", strings.Join(warnings, "\n"))
+	}
```

Comment:
We don't have many warnings but when we do they come from the `checkOptions()` call in each distro, which validates customizations for a given distro and image type.
The one warning I can point to (which I think is when warnings were introduced) is when a user is added in a blueprint for an ostree commit.  We wanted to deprecate that feature for RHEL 10, but now that edge is going away completely and bootc is a whole different thing, we should probably drop that.

Warnings are still probably useful for these kinds of things in the future though.

Reply:
Thanks, that is super useful! My worry about warnings like this is that if this runs in a CI/CD or a script many/most people will not see the warnings. I wonder if should error by default on warnings and add a `--warnings-okay` switch that would just put them on stderr (or of course the other way around, show them by default only and add `--strict/--Werror`. 

If we want to just show them right now I would tweak the code to:
```go
diff --git a/internal/manifestgen/manifestgen.go b/internal/manifestgen/manifestgen.go
index 4861658..a8b35ae 100644
--- a/internal/manifestgen/manifestgen.go
+++ b/internal/manifestgen/manifestgen.go
@@ -86,6 +86,7 @@ type (
 type Options struct {
        Cachedir          string
        Output            io.Writer
+       WarningsOutput    io.Writer
        Depsolver         DepsolveFunc
        ContainerResolver ContainerResolverFunc
        CommitResolver    CommitResolverFunc
@@ -96,6 +97,7 @@ type Options struct {
 type Generator struct {
        cacheDir string
        out      io.Writer
+       warn     io.Writer
 
        depsolver         DepsolveFunc
        containerResolver ContainerResolverFunc
@@ -114,6 +116,7 @@ func New(reporegistry *reporegistry.RepoRegistry, opts *Options) (*Generator, er
 
                cacheDir:          opts.Cachedir,
                out:               opts.Output,
+               warn:              opts.WarningsOutput,
                depsolver:         opts.Depsolver,
                containerResolver: opts.ContainerResolver,
                commitResolver:    opts.CommitResolver,
@@ -121,6 +124,9 @@ func New(reporegistry *reporegistry.RepoRegistry, opts *Options) (*Generator, er
        if mg.out == nil {
                mg.out = os.Stdout
        }
+       if mg.warn == nil {
+               mg.warn = os.Stderr
+       }
        if mg.depsolver == nil {
                mg.depsolver = defaultDepsolver
        }
@@ -150,9 +156,8 @@ func (mg *Generator) Generate(bp *blueprint.Blueprint, dist distro.Distro, imgTy
                return err
        }
        if len(warnings) > 0 {
-               // XXX: what can we do here? for things like json output?
-               // what are these warnings?
-               return fmt.Errorf("warnings during manifest creation: %v", strings.Join(warnings, "\n"))
+               // TODO: add --strict/--Werror option that errors here
+               fmt.Fprintf(mg.warn, "Warnings during manifest creation:\n%s", strings.Join(warnings, "\n"))
        }
        packageSpecs, _, err := mg.depsolver(mg.cacheDir, preManifest.GetPackageSetChains(), dist, a.Name())
        if err != nil {

```

My gut feeling for this is:
a) keep it strict and add --warnings-okay (name tbd)
b) do it in a followup to keep review focused

*But* I don't know how often those warnings happen in practice so maybe my gut-feeling is off and I'm happy to unblock this with the above diff too and add the "-Werror" later.

Reply:
There's also the FIPS warning: https://github.com/osbuild/images/blob/7600a5ae1c1e179e06d1cca3a97c7a464a6b9be7/pkg/distro/fedora/imagetype.go#L447-L450

Reply:
> I think a `-Werror` style option would be better but it depends what kind of use cases we're targeting and thinking about. You make a good point about warnings not being visible when the build is part of an automated pipeline, but I also think it'd be annoying for builds to break on a deprecation warning.
[..]

I guess it's hard to say which is better, from my PoV there are two cases:
1. the build was using something deprecated so from day-0 and the user is aware as the first run(s) where looked at, so if the deprecated thing stops working after say 2y that is no surprise
2. the build is using something that was fine on day-0 but we deprecated 1y in, no-one (probably) looks at the build logs because it's churning along fine for 1y. In 2y we remove the feature and the build fails, now people need to panic because their feature no longer works and is no longer available

Alternatively if we error by default it would be:
1. on day-0 the user puts in `--allow-warnings` no surprise here
2. he build is using something that was fine on day-0 but we deprecated 1y in, now it break. this is annoying and the user needs to change the job to use `--allow-warnings` (which we hopefully made clear from the error message). But they have 1y to plan how to replace the feature before it really goes away.

Of course this is all conjecture on my part, idk if people never look at their CI logs and I'm happy to just show warnings by default and add `-Werror` later (i.e. just apply the patch) [just wanted to eloborate on my gut feeling a little bit]

Reply:
You make a good argument there.  I think we need to think about it some more, but there's definitely some value for the cases you described.

We could even go as far as having a separate option for each warning (`--ignore-warning=fips-host`), so that users need to make explicit decisions about which things they need to worry about changing.